logictest: schema_change_in_txn failed with TransactionRetryError

pkg/sql/logictest/tests/local-legacy-schema-changer/local-legacy-schema-changer_test.TestLogic_schema_change_in_txn [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SqlLogicTestHighVModuleNightlyBazel/10428562?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_SqlLogicTestHighVModuleNightlyBazel/10428562?buildTab=artifacts#/) on release-23.1 @ [1a5dc2aad16ca4185afcda75d923804324dcfeb9](https://github.com/cockroachdb/cockroach/commits/1a5dc2aad16ca4185afcda75d923804324dcfeb9):


```
resident:                     8486912
abandoned_vm:                       0
extent_avail:                       0
                           n_lock_ops (#/sec)       n_waiting (#/sec)      n_spin_acq (#/sec)  n_owner_switch (#/sec)   total_wait_ns   (#/sec)     max_wait_ns  max_n_thds
large                              26       0               0       0               0       0              11       0               0         0               0           0
extent_avail                       30       0               0       0               0       0              13       0               0         0               0           0
extents_dirty                      29       0               0       0               0       0              13       0               0         0               0           0
extents_muzzy                      26       0               0       0               0       0              11       0               0         0               0           0
extents_retained                   30       0               0       0               0       0              13       0               0         0               0           0
decay_dirty                        28       0               0       0               0       0              12       0               0         0               0           0
decay_muzzy                        26       0               0       0               0       0              11       0               0         0               0           0
base                               55       0               0       0               0       0              12       0               0         0               0           0
tcache_list                        26       0               0       0               0       0              11       0               0         0               0           0
hpa_shard                           0       0               0       0               0       0               0       0               0         0               0           0
hpa_shard_grow                      0       0               0       0               0       0               0       0               0         0               0           0
hpa_sec                             0       0               0       0               0       0               0       0               0         0               0           0
bins:           size ind    allocated      nmalloc (#/sec)      ndalloc (#/sec)    nrequests   (#/sec)  nshards      curregs     curslabs  nonfull_slabs regs pgs   util       nfills (#/sec)     nflushes (#/sec)       nslabs     nreslabs (#/sec)      n_lock_ops (#/sec)       n_waiting (#/sec)      n_spin_acq (#/sec)  n_owner_switch (#/sec)   total_wait_ns   (#/sec)     max_wait_ns  max_n_thds
                     ---
large:          size ind    allocated      nmalloc (#/sec)      ndalloc (#/sec)    nrequests (#/sec)  curlextents
                     ---
             8388608  72      8388608            2       0            1       0            2       0            1
                     ---
extents:        size ind       ndirty        dirty       nmuzzy        muzzy    nretained     retained       ntotal        total
                     ---
             2097152  31            0            0            0            0            1      2093056            1      2093056
                     ---
Bytes in small extent cache: 0
HPA shard stats:
  Purge passes: 0 (0 / sec)
  Purges: 0 (0 / sec)
  Hugeifies: 0 (0 / sec)
  Dehugifies: 0 (0 / sec)

  In full slabs:
      npageslabs: 0 huge, 0 nonhuge
      nactive: 0 huge, 0 nonhuge 
      ndirty: 0 huge, 0 nonhuge 
      nretained: 0 huge, 0 nonhuge 
  In empty slabs:
      npageslabs: 0 huge, 0 nonhuge
      nactive: 0 huge, 0 nonhuge 
      ndirty: 0 huge, 0 nonhuge 
      nretained: 0 huge, 0 nonhuge 

                size ind npageslabs_huge    nactive_huge     ndirty_huge  npageslabs_nonhuge     nactive_nonhuge      ndirty_nonhuge   nretained_nonhuge
                     ---
--- End jemalloc statistics ---
[05:20:19] --- total progress: 191 statements
--- total: 191 tests, 2 failures
    --- FAIL: TestLogic_schema_change_in_txn/rollback_mutations (0.76s)
```
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestLogic_schema_change_in_txn.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-28541

--- a/pkg/ccl/logictestccl/testdata/logic_test/schema_change_in_txn
+++ b/pkg/ccl/logictestccl/testdata/logic_test/schema_change_in_txn
@@ -1,3 +1,7 @@
+# Skip the rest of the test if a retry occurs. They can happen and are fine
+# but there's no way to encapsulate that in logictests.
+skip_on_retry
+
 # Backing up and restoring a descriptor will increment the version of the
 # descriptor before restoring it so we cannot achieve the expected behaviour in
 # this test.

--- a/pkg/sql/logictest/logic.go
+++ b/pkg/sql/logictest/logic.go
@@ -3130,8 +3130,8 @@ func (t *logicTest) maybeSkipOnRetry(err error) {
 func (t *logicTest) verifyError(
 	sql, pos, expectNotice, expectErr, expectErrCode string, err error,
 ) (bool, error) {
+	t.maybeSkipOnRetry(err)
 	if expectErr == "" && expectErrCode == "" && err != nil {
-		t.maybeSkipOnRetry(err)
 		return t.unexpectedError(sql, pos, err)
 	}
 	if expectNotice != "" {

